core/src/main/java/org/eclipse/hono/util/MessageDigestPasswordEncoder.java

```diff
@@ -0,0 +1,64 @@
+/*******************************************************************************
+ * Copyright (c) 2018 Red Hat Inc and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Red Hat - initial creation
+ *******************************************************************************/
+package org.eclipse.hono.util;
+
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.Base64;
+
+/**
+ * Implementation of {@link PasswordEncoder} that properly uses salt.
+ */
+public class MessageDigestPasswordEncoder implements PasswordEncoder {
+
+    private final MessageDigest messageDigest;
+
+    /**
+     * Create message digest password encoder with specified hash function.
+     *
+     * @param hashFunction - hash function to be used
+     * @throws IllegalArgumentException if hash function is not valid
+     */
+    public MessageDigestPasswordEncoder(final String hashFunction) {
+        try {
+            messageDigest = MessageDigest.getInstance(hashFunction);
+        } catch (NoSuchAlgorithmException e) {
+            throw new IllegalArgumentException("No such hash function: " + e);
+        }
+    }
+
+    @Override
+    public String encode(final CharSequence rawPassword) {
```

Comment:
IMHO we should also add a JavaDoc to this method which describes that the returned hash is created as defined by Hono's Credentials API.